ProcessLib/ThermoMechanicalPhaseField/ThermoMechanicalPhaseFieldFEM-impl.h

```diff
@@ -0,0 +1,328 @@
+    auto const& local_u =
+        local_coupled_solutions.local_coupled_xs[_mechanics_related_process_id];
+    auto const& local_T =
+        local_coupled_solutions.local_coupled_xs[_heat_conduction_process_id];
+    assert(local_T.size() == temperature_size);
+    assert(local_d.size() == phasefield_size);
+    assert(local_u.size() == displacement_size);
+
+    auto d = Eigen::Map<
+        typename ShapeMatricesType::template VectorType<phasefield_size> const>(
+        local_d.data(), phasefield_size);
+
+    auto u = Eigen::Map<typename ShapeMatricesType::template VectorType<
+        displacement_size> const>(local_u.data(), displacement_size);
+
+    auto T = Eigen::Map<typename ShapeMatricesType::template VectorType<
+        temperature_size> const>(local_T.data(), temperature_size);
+
+    auto local_Jac = MathLib::createZeroedMatrix<
+        typename ShapeMatricesType::template MatrixType<
+            displacement_size, displacement_size>>(
+        local_Jac_data, displacement_size, displacement_size);
+
+    auto local_rhs =
+        MathLib::createZeroedVector<typename ShapeMatricesType::
+                                        template VectorType<displacement_size>>(
+            local_b_data, displacement_size);
+
+    double const& dt = _process_data.dt;
+
+    SpatialPosition x_position;
+    x_position.setElementID(_element.getID());
+
+    int const n_integration_points = _integration_method.getNumberOfPoints();
+    for (int ip = 0; ip < n_integration_points; ip++)
+    {
+        x_position.setIntegrationPoint(ip);
+        auto const& w = _ip_data[ip].integration_weight;
+        auto const& N = _ip_data[ip].N;
+        auto const& dNdx = _ip_data[ip].dNdx;
+
+        auto const x_coord =
+            interpolateXCoordinate<ShapeFunction, ShapeMatricesType>(_element,
+                                                                     N);
+        auto const& B =
+            LinearBMatrix::computeBMatrix<DisplacementDim,
+                                          ShapeFunction::NPOINTS,
+                                          typename BMatricesType::BMatrixType>(
+                dNdx, N, x_coord, _is_axially_symmetric);
+
+        auto& eps = _ip_data[ip].eps;
+
+        auto const& C_tensile = _ip_data[ip].C_tensile;
+        auto const& C_compressive = _ip_data[ip].C_compressive;
+
+        auto const& sigma = _ip_data[ip].sigma;
+
+        auto const k = _process_data.residual_stiffness(t, x_position)[0];
+        auto rho_sr = _process_data.solid_density(t, x_position)[0];
+        auto const alpha = _process_data.linear_thermal_expansion_coefficient(
+            t, x_position)[0];
+        double const T0 = _process_data.reference_temperature;
+        auto const& b = _process_data.specific_body_force;
+
+        double const T_ip = N.dot(T);
+        double const delta_T = T_ip - T0;
+        // calculate real density
+        double const rho_s = rho_sr / (1 + 3 * alpha * delta_T);
```

Comment:
See my comment for another rho_s

Reply:
Correct it or use rho_sr as rho_s directly. For solid of geomaterials or metal, the factor 1 / (1 +3 * alpha * dT) to their density are tiny.